### ISSUE TITLE
cmake on clang32 reports wrong configuration time

### ISSUE THREAD

**Issue by madebr** [msys2/MINGW-packages#17795]
### Description / Steps to reproduce the issue

Recent cmake releases reports the time it takes to configure a project in seconds at the end.
It looks like on clang32, this is somehow broken.

For example, see this output of a GitHub workflow:
https://github.com/libsdl-org/SDL/actions/runs/5589750584/jobs/10218501241#step:9:271

### Expected behavior

```
 -- Configuring done (32.9s)
-- Generating done (0.1s)
-- Build files have been written to: D:/a/SDL/SDL/build
```

### Actual behavior

```
 -- Configuring done (-9123292659426253850436719799017442136586774209552161337536467353746771162918819034520641703491042846408815082120540550597342102467005169222967783506967346641314978059609663676974096259599798996998937607570840069391076737937496429167352331447416802841585171636136443462884803862862218511601527855135072452608.0s)
-- Generating done (-0.0s)
-- Build files have been written to: D:/a/SDL/SDL/build
```

### Verification

- [X] I have verified that my MSYS2 is up-to-date before submitting the report (see https://www.msys2.org/docs/updating/)

### Windows Version

windows-latest GitHub runner

### MINGW environments affected

- [ ] MINGW64
- [ ] MINGW32
- [ ] UCRT64
- [ ] CLANG64
- [X] CLANG32
- [ ] CLANGARM64

### Are you willing to submit a PR?

I have no experience with clang :(

**Comment by Biswa96** [comment]
Thank you for reporting this issue. I have reproduced the issue with the following example. The code is based on the commit which calculates the time difference after cmake configure https://github.com/Kitware/CMake/commit/5f0c5ec49b553cbe4482d9a8cf7e8889dc90e759

```cpp
#include <chrono>
#include <iomanip>
#include <iostream>
#include <thread>

int main()
{
    auto startTime = std::chrono::steady_clock::now();
    std::this_thread::sleep_for(std::chrono::milliseconds(3000));
    auto endTime = std::chrono::steady_clock::now();
    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
    std::cout << "Configuring done (" << std::fixed << std::setprecision(1) << ms.count() / 1000.0L << "s)\n";
}
```

* UCRT64:
```
$ g++ test.cpp && ./a.exe && echo $?
Configuring done (3.0s)
0
```

* CLANG64

```
$ clang++ test.cpp && ./a.exe && echo $?
Configuring done (0.0s)
0
```

* CLANG32:

```
$ clang++ test.cpp && ./a.exe && echo $?
Configuring done (-21495.8s)
0
```

**Comment by mstorsjo** [comment]
This seems to be a case with printf formatting of long doubles, which requires `__USE_MINGW_ANSI_STDIO=1` to be defined. Adding that flag to libc++ when building it (e.g. adding it to `CMAKE_CXX_FLAGS`) seems to help. If added to `LIBCXX_EXTRA_SITE_DEFINES`, e.g. as `LIBCXX_EXTRA_SITE_DEFINES=__USE_MINGW_ANSI_STDIO=1`, the same define also should get set in libcxx's headers, affecting the user's code after getting it included. (Libstdc++ does this too, fwiw.) But for this particular case, just setting it within libc++ itself seems to be enough.

**Comment by mstorsjo** [comment]
Also, note that in the calculation `ms.count() / 1000.0L`, `1000.0L` is a `long double` constant. If the trailing `L` is removed, this becomes a regular double, and the formatting works correctly regardless of `__USE_MINGW_ANSI_STDIO`.

**Comment by jeremyd2019** [comment]
> This seems to be a case with printf formatting of long doubles, which requires `__USE_MINGW_ANSI_STDIO=1` to be defined. Adding that flag to libc++ when building it (e.g. adding it to `CMAKE_CXX_FLAGS`) seems to help.

I thought that define was set by default for builds here.  In makepkg_mingw.conf has:
```
  CPPFLAGS="-D__USE_MINGW_ANSI_STDIO=1"
```

for all MSYSTEMs.  Perhaps it needs to be passed to some other option (like `CMAKE_CXX_FLAGS`) while building libc++?

**Comment by mstorsjo** [comment]
> > This seems to be a case with printf formatting of long doubles, which requires `__USE_MINGW_ANSI_STDIO=1` to be defined. Adding that flag to libc++ when building it (e.g. adding it to `CMAKE_CXX_FLAGS`) seems to help.
> 
> I thought that define was set by default for builds here. In makepkg_mingw.conf has:
> 
> ```
>   CPPFLAGS="-D__USE_MINGW_ANSI_STDIO=1"
> ```
> 
> for all MSYSTEMs. Perhaps it needs to be passed to some other option (like `CMAKE_CXX_FLAGS`) while building libc++?

Yes, I'm not sure if `CPPFLAGS` is picked up automatically by cmake. Some variables are (like `CC` and maybe `CFLAGS`), but I'm not sure about `CPPFLAGS`.

**Comment by mati865** [comment]
Indeed looks like CPPFLAGS are ignored by CMake: https://gitlab.kitware.com/cmake/cmake/-/issues/12928

**Comment by lazka** [comment]
Confirmed that this is fixed via #21796
